### DISCUSSION TITLE
How to take the time average using time bounds?

### DISCUSSION THREAD

**Discussion by briochemc** [pydata/xarray#7580] [General]
👋 Noob here so maybe this is a dumb question but here I go. 

Given a dataset `ds` with some variable `foo` with a `time` coordinate (among others like `x`, `y`, `z`, and so on) that also includes `time_bnds`, what's the preferred way to get the time-weighted average of `foo` (that uses the `time_bnds` data)? 

Adapting [*Calculating Seasonal Averages from Time Series of Monthly Means* from the docs](https://docs.xarray.dev/en/stable/examples/monthly-means.html#Calculating-Seasonal-Averages-from-Time-Series-of-Monthly-Means), I know I could manually extract `deltatime = ds["time_bnds"].sel(bnds=1) -  ds["time_bnds"].sel(bnds=0)` and build weights from `weights = deltatime / deltatime.sum()` and then do the weighted sum `(foo * weights).sum()` or something similar, but I was hoping there is a predefined automatic way to that in a single liner with xarray. Is that the case?

**Comment by arfriedman** [comment]
Hi @briochemc check out [xCDAT](https://xcdat.readthedocs.io/en/stable/index.html), which builds on xarray.
In particular: https://xcdat.readthedocs.io/en/stable/examples/temporal-average.html

**Reply by briochemc** [reply]
Hey! Thanks for the pointer @arfriedman! However, the part of the docs linked does not mention the time bounds and if they are used or not... Am I missing something?

**Reply by arfriedman** [reply]
hi @briochemc here's information on the xCDAT temporal accessor, which uses time bounds: https://xcdat.readthedocs.io/en/stable/generated/xcdat.temporal.TemporalAccessor.html
Note that the functionality is still evolving: https://github.com/xCDAT/xcdat/issues/412

**Reply by tomvothecoder** [reply]
Thanks @arfriedman for plugging xCDAT! Yes the temporal accessor uses time bounds.

@briochemc, we'll update the [docs](https://github.com/xCDAT/xcdat/issues/426) to make this info clear. Let us know if you have any questions over at https://github.com/xCDAT/xcdat/discussions!

**Reply by tomvothecoder** [reply]
Below are the sections of code in xCDAT where time bounds are used to generate weights and where weights are applied to the variable:
  - `_averager()`: https://github.com/xCDAT/xcdat/blob/e6dbed5c7037a0ad09a90bb364f7aee3a25915e3/xcdat/temporal.py#L701-L708
  - `_average()`: https://github.com/xCDAT/xcdat/blob/e6dbed5c7037a0ad09a90bb364f7aee3a25915e3/xcdat/temporal.py#L1015-L1018
  - `_group_average()`: https://github.com/xCDAT/xcdat/blob/e6dbed5c7037a0ad09a90bb364f7aee3a25915e3/xcdat/temporal.py#L1049-L1067
  - `_get_weights()` : https://github.com/xCDAT/xcdat/blob/e6dbed5c7037a0ad09a90bb364f7aee3a25915e3/xcdat/temporal.py#L1090-L1148

**Reply by briochemc** [reply]
BTW, I just realized... Should you not add xcdat to the [ecosystem list](https://docs.xarray.dev/en/stable/ecosystem.html)?

**Reply by tomvothecoder** [reply]
@briochemc It was recently added to the `main` branch and will be included in the next release.
"Latest" version of docs: https://docs.xarray.dev/en/latest/ecosystem.html#geosciences
